paseos/activities/activity_processor.py

```diff
@@ -0,0 +1,110 @@
+            power_consumption_in_watt >= 0
+        ), "Power consumption has to be positive but was specified as " + str(
+            power_consumption_in_watt
+        )
+
+        self._power_consumption_in_watt = power_consumption_in_watt
+        self.update_interval = update_interval
+        self.is_started = False
+        self._task = None
+        self._paseos_instance = paseos_instance
+        self._activity_runner = activity_runner
+        self._advance_paseos_clock = advance_paseos_clock
+
+    async def start(self):
+        """Starts the processor."""
+        if not self.is_started:
+            logger.trace("Starting ActivityProcessor.")
+            # Remember when we start
+            self.start_time = timer()
+            self.is_started = True
+            # Start task to call func periodically:
+            self._task = asyncio.ensure_future(self._run())
+
+    async def stop(self):
+        """Stops the processor."""
+        if self.is_started:
+            logger.trace("Stopping ActivityProcessor.")
+            # Calculate elapsed time since last update
+            elapsed_time = timer() - self.start_time
+            # Perform final update
+            await self._update(elapsed_time)
+            self.is_started = False
+            # Stop task and await it stopped:
+            self._task.cancel()
+            with suppress(asyncio.CancelledError):
+                await self._task
+
+    async def _update(self, elapsed_time: float):
+        """Updates the processor and optionally local actor.
+
+        Args:
+            elapsed_time (float): Elapsed time in seconds.
+        """
+        assert elapsed_time > 0, "Elapsed time cannot be negative."
+        logger.debug("Running ActivityProcessor update.")
+        logger.debug(f"Time since last update: {elapsed_time}s")
+        if self._advance_paseos_clock:
+            self._paseos_instance.advance_time(elapsed_time)
+
+        self._paseos_instance.local_actor.discharge(
+            self._power_consumption_in_watt, elapsed_time
+        )
+
+    async def _run(self):
+        """Main processor loop. Will track time, update paseos and check constraints of the activity."""
+        while True:
+            # Calculate elapsed time since last update
+            elapsed_time = timer() - self.start_time
+
+            # Make sure we don't update more frequently than specified
+            # by waiting till we at least update_interval time elapsed.
+            if elapsed_time < self.update_interval:
+                await asyncio.sleep(self.update_interval - elapsed_time)
+                elapsed_time = timer() - self.start_time
+
+            # Start new timer
+            self.start_time = timer()
+
+            # Perform the update
+            await self._update(elapsed_time)
+
+            # Check if the activity should still run
+            # otherwise stop it and then the processor
+            if self._activity_runner.has_constraint():
```

Comment:
if activity_runner is list, we need to check multiple constraints.

Reply:
see above